### ISSUE TITLE
False positive for MSF_MUTABLE_SERVLET_FIELD

### ISSUE THREAD

**Issue by guillerodriguez** [spotbugs/spotbugs#2131]
Spotbugs is reporting MSF_MUTABLE_SERVLET_FIELD in the following test case:

```java
import java.io.IOException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

public class MutableFieldServlet extends HttpServlet
{
    private String etag;

    @Override
    public void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException
    {
        synchronized (this)
        {
            if (etag == null)
                etag = "\"" + System.currentTimeMillis() + "\"";
        }

        resp.addHeader("ETag", etag);
        resp.getWriter().println("Hello world");
    }
}
```

However, there should be no race condition here:

- Writes to the mutable field are only done in the synchronized block
- The field is read without synchronization, however the visibility guarantees of synchronized blocks guarantee that all threads that do not initialise `etag` will already see an initialised value after they exit the synchronized block.

This is reproducible with 4.7.1.

**Comment by ThrawnCA** [comment]
The intent of MSF_MUTABLE_SERVLET_FIELD is to entirely discourage the use of mutable fields in servlets.

In this case, although you have addressed the possibility of race conditions, it might be better for the field to be `static`, since it's supposed to be initialised just once. Or perhaps a `final` field initialised immediately on server start. So, the "don't use mutable servlet fields" advice is still valid.

(I'm also dubious about this being a suitable value for an [ETag](https://en.wikipedia.org/wiki/HTTP_ETag). ETags are meant to work like a fingerprint for the resource data, for the purpose of efficiently caching, so a hash would make a lot more sense than a timestamp. But that's a separate issue.)

**Comment by guillerodriguez** [comment]
> In this case, although you have addressed the possibility of race conditions, it might be better for the field to be `static`, since it's supposed to be initialised just once. Or perhaps a `final` field initialised immediately on server start. So, the "don't use mutable servlet fields" advice is still valid.

I agree that "don't use mutable server fields" is valid as general advice, but the MSF_MUTABLE_SERVLET_FIELD warning is not always raised:
 * If I include the second field access in the synchronized block, the warning goes away
 * If I declare the variable as volatile, the warning goes away

I assume this means there is already some logic to check whether there are potential race conditions, and to only raise the warning in that case. That's why I interpreted this to be a false positive.

> 
> (I'm also dubious about this being a suitable value for an [ETag](https://en.wikipedia.org/wiki/HTTP_ETag). ETags are meant to work like a fingerprint for the resource data, for the purpose of efficiently caching, so a hash would make a lot more sense than a timestamp. But that's a separate issue.)

Yes, of course. The original code uses a hash. This is just a simplified test case.
